File: core/src/main/java/com/hgianastasio/biblioulbrav2/core/base/UseCase.java
package com.hgianastasio.biblioulbrav2.core.base;


import com.hgianastasio.biblioulbrav2.core.base.boundaries.OnErrorListener;
import com.hgianastasio.biblioulbrav2.core.base.boundaries.OnResultListener;

import java.util.concurrent.ThreadPoolExecutor;

/**
 * Created by heitorgianastasio on 4/24/17.
 */

public abstract class UseCase<T,P> {

    protected ThreadPoolExecutor executor;

    protected UseCase(ThreadPoolExecutor executor){
        this.executor = executor;
    }

    protected abstract void process(P params, OnResultListener<T> resultListener, OnErrorListener errorListener);

    public void execute(P params, OnResultListener<T> resultListener, OnErrorListener errorListener){
        executor.execute(()->process(params,resultListener,errorListener));
    }





}

File: core/src/main/java/com/hgianastasio/biblioulbrav2/core/base/UseCase.kt
package com.hgianastasio.biblioulbrav2.core.base

import com.hgianastasio.biblioulbrav2.core.base.boundaries.OnErrorListener
import com.hgianastasio.biblioulbrav2.core.base.boundaries.OnResultListener
import java.util.concurrent.ThreadPoolExecutor

/**
 * Created by heitorgianastasio on 4/24/17.
 */
abstract class UseCase<T, P> protected constructor(protected var executor: ThreadPoolExecutor) {
    protected abstract fun process(params: P, resultListener: OnResultListener<T>, errorListener: OnErrorListener)
    fun execute(params: P, resultListener: OnResultListener<T>, errorListener: OnErrorListener) {
        executor.execute { process(params, resultListener, errorListener) }
    }

}